pkg/collector/corechecks/oracle/oracle.go

```diff
@@ -0,0 +1,214 @@
+	SQLIDs map[string]int
+	//ForceMatchingSignatures map[uint64]int
+	ForceMatchingSignatures map[string]int
+}
+
+type Check struct {
+	core.CheckBase
+	config                                  *config.CheckConfig
+	db                                      *sqlx.DB
+	hostname                                string
+	dbmEnabled                              bool
+	agentVersion                            string
+	checkInterval                           float64
+	tags                                    []string
+	cdbName                                 string
+	statementsFilter                        StatementsFilter
+	statementMetricsMonotonicCountsPrevious map[StatementMetricsKeyDB]StatementMetricsMonotonicCountDB
+	dbHostname                              string
+	dbVersion                               string
+	driver                                  string
+}
+
+// Run executes the check.
+func (c *Check) Run() error {
+	if c.db == nil {
+		db, err := c.Connect()
+		if err != nil {
+			c.Teardown()
+			return err
+		}
+		c.db = db
+	}
+
+	if c.hostname == "" {
+		hostname, err := os.Hostname()
+		if err != nil {
+			return fmt.Errorf("failed to get hostname: %w", err)
+		}
+		c.hostname = hostname
+	}
+
+	if c.dbmEnabled {
+		err := c.SampleSession()
+		if err != nil {
+			return err
+		}
+
+		_, err = c.StatementMetrics()
+		if err != nil {
+			return err
+		}
+
+	}
+
+	return nil
+}
+
+// Connect establishes a connection to an Oracle instance and returns an open connection to the database.
+func (c *Check) Connect() (*sqlx.DB, error) {
+
+	var connStr string
+	var oracleDriver string
+	if c.config.TnsAlias != "" {
+		connStr = fmt.Sprintf(`user="%s" password="%s" connectString="%s"`, c.config.Username, c.config.Password, c.config.TnsAlias)
+		oracleDriver = "godror"
+	} else {
+		//godror ezconnect string
+		if c.config.InstanceConfig.InstantClient {
+			oracleDriver = "godror"
+			connStr = fmt.Sprintf("%s/%s@%s/%s", c.config.Username, c.config.Password, c.config.Server, c.config.ServiceName)
+		} else {
+			oracleDriver = "oracle"
+			connStr = go_ora.BuildUrl(c.config.Server, c.config.Port, c.config.ServiceName, c.config.Username, c.config.Password, map[string]string{})
+			// https://github.com/jmoiron/sqlx/issues/854#issuecomment-1504070464
+			sqlx.BindDriver("oracle", sqlx.NAMED)
+		}
+	}
+	c.driver = oracleDriver
+
+	log.Infof("driver: %s, Connect string: %s", oracleDriver, connStr)
+
+	db, err := sqlx.Open(oracleDriver, connStr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to oracle instance: %w", err)
+	}
+	err = db.Ping()
+	if err != nil {
+		return nil, fmt.Errorf("failed to ping oracle instance: %w", err)
+	}
+
+	db.SetMaxOpenConns(10)
+
+	if c.cdbName == "" {
+		row := db.QueryRow("SELECT name FROM v$database")
+		err = row.Scan(&c.cdbName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to query db name: %w", err)
+		}
+	}
+	c.tags = append(c.tags, fmt.Sprintf("cdb:%s", c.cdbName))
```

Comment:
If `Connect` returns an error after this, next run we append the tag again, which may be different, and a memory leak either way. Would it be possible to make `Connect` safe to be called multiple times?

Reply:
Corrected it: now the tags are appended only in the case they weren't previously appended.